tests/stack/test_api.py

```diff
@@ -338,3 +338,65 @@ def test_deploy_resource(self, stack_api):
         }
         with pytest.raises(StackError):
             stack_api._deploy_resource(resource_badtype)
+
+    def test_download_workspace(self, stack_api, tmpdir):
+        """
+            stack_api._deploy_workspace should call certain workspace client functions depending
+            on object_type and error when object_type is defined incorrectly.
+        """
+        test_deploy_output = {'test': 'test'}  # default deploy_output return value
+
+        stack_api.workspace_client.client = mock.MagicMock()
+        stack_api.workspace_client.client.get_status.return_value = test_deploy_output
+        stack_api.workspace_client.export_workspace = mock.MagicMock()
+        stack_api.workspace_client.export_workspace_dir = mock.MagicMock()
+
+        test_workspace_nb_properties = TEST_WORKSPACE_NB_PROPERTIES.copy()
+        test_workspace_nb_properties.update(
+            {'source_path': os.path.join(tmpdir.strpath,
+                                         test_workspace_nb_properties['source_path'])})
+        test_workspace_dir_properties = TEST_WORKSPACE_DIR_PROPERTIES.copy()
+        test_workspace_dir_properties.update(
+            {'source_path': os.path.join(tmpdir.strpath,
+                                         test_workspace_dir_properties['source_path'])})
+
+        stack_api._download_workspace(test_workspace_dir_properties, True)
```

Comment:
As a general comment, it is best not to test private methods because private methods are flexible to change. Rather, you should just test the public methods instead.

Reply:
oh man i have a lot of tests that test private methods... I'll note this for future developments.

Reply:
I agree the tests should focus on APIs. However, I feel sometimes it is good to test helper (private) method as long as the method perform a well defined task.